### PR TITLE
[Az.RecoveryServices.Backup] Added support for Private Endpoints and Immutable Vaults

### PR DESCRIPTION
**Pull request by hiaga**
## Description

Added support for enable/disable Public Network Access and PrivateEndpoints
Added support for Immutable Vaults
Added List RP expiry time

Added support for RetainRecoveryPointsAsPerPolicy in Disable-AzRecoveryServicesBackupProtection cmdlet. Now user can suspend backups and retain RPs as per policy

Added RecoveryServices, RecoveryServices.Backup, RecoveryServices.Backup.CrossRegionRestore management SDK

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by isra-fel** [comment]
Please change the base branch to release-2023-02-07 if the PR targets Feb release.

File: src/RecoveryServices/RecoveryServices.Backup.Helpers/Conversions/RecoveryPointConversions.cs [hunk 1]
```diff
@@ -775,6 +794,12 @@ public static ServiceClientModel.RecoveryPointTierType GetServiceClientRecoveryP
                 }
             }
 
+            /*if (recoveryPoint.RecoveryPointProperties != null)
```
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices.Backup.Helpers/Conversions/RecoveryPointConversions.cs:797]: Please remove commented code unless you'll need it in the future, in which case add a note of when to uncomment

File: src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Item/DisableAzureRmRecoveryServicesBackupProtection.cs [hunk 2]
```diff
@@ -54,6 +54,14 @@ public SwitchParameter RemoveRecoveryPoints
             set { DeleteBackupData = value; }
         }
 
+        /// <summary>
+        /// If this option is used, all the data backed up for this item will 
+        /// expire as per the protection policy retention settings
+        /// </summary>
+        [Parameter(Mandatory = false,  HelpMessage = ParamHelpMsgs.Item.SuspendBackupOption)]
+        [ValidateNotNullOrEmpty]
```
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Item/DisableAzureRmRecoveryServicesBackupProtection.cs:62]:
```suggestion
```
Redundant for a switch parameter

File: src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Item/DisableAzureRmRecoveryServicesBackupProtection.cs [hunk 3]
```diff
@@ -90,6 +98,11 @@ public override void ExecuteCmdlet()
                             throw new ArgumentException(String.Format(Resources.DisableWithRetainBackupNotCrititcal));
                         }
 
+                        if(DeleteBackupData && RetainRecoveryPointsAsPerPolicy.IsPresent)
+                        {
+                            throw new ArgumentException(String.Format(Resources.CantRemoveAndRetainRPsSimultaneously));
```
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Item/DisableAzureRmRecoveryServicesBackupProtection.cs:103]: Please use the ["AzPS" series of exceptions](https://github.com/Azure/azure-powershell-common/tree/main/src/Common/Exceptions) whenever possible. It helps with detailed telemetry data.

**Review by isra-fel — CHANGES_REQUESTED** [review]

File: src/RecoveryServices/RecoveryServices.Backup.Helpers/Conversions/RecoveryPointConversions.cs (hunk 1 above)
Comment (hiaga) [reply src/RecoveryServices/RecoveryServices.Backup.Helpers/Conversions/RecoveryPointConversions.cs:797]: will need in future, let me add a note

**Review by isra-fel — APPROVED** [review]

**Comment by NThompson480** [comment]
I am sorry if this is not the proper way to address this issue, I am happy to do so if there is a better way.

I have been trying to use the RetainRecoveryPointsAsPerPolicy switch and have been getting the following error:

Disable-AzRecoveryServicesBackupProtection: The Vault and Bakcup Item is in Immutable state where destructive operations are not supported. The Vault and Backup Item is immutable, hence operation is denied. Refer to the documentation to learn more: https://aka.ms/AB-ImmutableVaults-States 

I started looking into the code, trying to understand what the DisableProtectionWithDeleteData command does when the RetainRecoveryPointsAsPerPolicy switch is used, when the RemoveRecoveryPoints switch is used, and when no switch is used. I found that one of three methods are called SuspendBackup, DisableProtectionWithDeleteData, and DisableProtection.

So with the switch I am having issues with, RetainRecoveryPointsAsPerPolicy, it would suspend backups and retain the recovery points in alignment with the policy. The RemoveRecoveryPoints switch is obvious. And then, no switch... that is what I would stop the VM backup and it also appears to retain the recovery points.

If I am deleting a VM, but I want to retain the recovery points per the retention (backup) policy, which switch would I use?

When you 'Stop Backups' from the portal, you have two options: 'Retain backup data' and 'Delete backup data'. Delete backup data is obviously equivalent to using the RemoveRecoveryPoints switch. Is Retain backup data equivalent to RetainRecoveryPointsAsPerPolicy or no switch?

**Comment by vidai-msft** [comment]
@NThompson480 can you please raise the question on the GitHub issues? We will triage and then help redirect to the concerned team. Thanks!

**Pull request merged**